include/dev/usb.h

```diff
@@ -145,4 +149,24 @@ typedef struct usb_endpoint_descriptor {
 #define UE_BULK 0x02
 #define UE_INTERRUPT 0x03
 
+/*
+ * Implementation specific constructs.
+ */
+
+typedef enum {
+  TF_CONTROL = 1,   /* control transfer */
+  TF_BULK = 2,      /* bulk transfer */
+  TF_INTERRUPT = 4, /* interrupt transfer */
+  TF_STALLED = 8,   /* STALL condition encountered */
+  TF_ERROR = 16,    /* errors other than STALL */
+} usb_transfer_flags_t;
+
+/* USB buffer used for USB transfers. */
+typedef struct usb_buf {
+  ringbuf_t buf;              /* write source or read destination */
+  condvar_t cv;               /* wait for the transfer to complete */
+  spin_t lock;                /* buffer guard */
+  usb_transfer_flags_t flags; /* transfer description */
+} usb_buf_t;
```

Comment:
Do you need that `flags` field to be there? I have some mixed feelings about it because of the transfer type flags.

There are two procedures for performing transfers and as far as I understand the `usbhc_read` and `usbhc_write` are used for `TF_BULK` and `TF_INTERRUPT` transfers, while `usb_control_read` and `usb_control_write` are used for `TF_CONTROL` transfers, meaning that the flags need to match the used procedure. However, I've been wondering wether this is really necessary,. Perhaps you could get rid of `TF_CONTROL` and make it implicit when using `usbhc_control_read` and `usb_control_write`. `TF_BULK` and `TF_INTERRUPT` then could be thrown out of the `usb_transfer_flags_t` and used as an enum parameter for `usbhc_read` and `usbhc_write` procedures. This would remove the need to ensure setting the correct transfer type when constructing a usb transfer buffer.

Reply:
Transfer flags are gone. The only thing I need now is error flags.